docs/proposals/20191017-kubeadm-based-control-plane.md

```diff
@@ -381,23 +408,46 @@ spec:
 
 * Infrastructure templates are expected to be immutable, so infrastructure template contents do not have to hashed in order to detect
   changes.
-   
+
+##### KubeadmControlPlane rollout (using delete-and-create)
+
+KCP rollout delete-and-create is recommended for resource constrained environment like bare-metal, OpenStack or vSphere resource pools, etc.
+
+- Triggered by:
+    - Changes to Version
+    - Changes to the kubeadmConfigSpec
+    - Changes to the infrastructureRef
+    - `KubeadmControlPlane.Spec.UpgradeStrategy` is set
+
+- Rollout operations rely on scale up and scale down which are be blocked based on Etcd and control plane health checks.
+
+- The algorithm is the following:
+  - Verify control plane replica count is >= 3
```

Comment:
How do we think to make it visible to the user the fact that the rollout can be performed due to this reason?

Reply:
I did a little rethinking with the logic here. I think we don't need to be worried about the replica counts at all. Since current implementation does not even support the deployment of KCP with even replica count and returns on error:

`The KubeadmControlPlane "controlplane-name" is invalid: spec.replicas: Forbidden: cannot be an even number when using managed etcd`

Also KCP does not upgrade if replica count is for example 2 we can rely on the fact that replica count is uneven when user is trying to upgrade. So I would say we only make sure that we scale up when replica count is 1 and `KubeadmControlPlane.Spec.UpgradeStrategy` is set. Any thoughts?

Reply:
IMO we should have an explicit rule checking that *current* replicas >=3.

The main reason for having this explicit rule is that even do it is not allowed to specify an even number *desired* replica count, then in the current KCP implementation the rollout logic takes precedence on creating the missing machines, and as a result, it could happen that:
- Set desired replica count = 3
- Create first machine (desired replicas = 3, current replicas =1)
- Change machine spec
- KCP start to rollout existing machines, stopping to provision the missing ones 
- if UpgradeStrategy == scale-in, without the explicit rule above, the only existing replica will be deleted 😞 

So the rule is ok for me, might be worth to specify *current* machines; it is only important to inform the user when this rule is preventing rollout to happen

Reply:
That is ok for me. Do we have any preferences how we should inform the user?

Reply:
Should we also prevent users to do initial deployment of the KCP with: 
- UpgradeStrategy == scale-in 
- replica count < 3 

For this we could add something as follows into `kubeadm_control_plane_webhook.go` :

```
	if in.Spec.UpgradeStrategy.Type == "ScaleIn" && *in.Spec.Replicas < 3 {
		allErrs = append(
			allErrs,
			field.Forbidden(
				field.NewPath("spec", "replicas"),
				"cannot set scaleIn rollout stragegy with less than three initial replicas",
			),
		)
	} 
```

Reply:
> Do we have any preferences how we should inform the user?

We are converging on using conditions for all the user-facing messages, so this could a warning on the `MachinesSpecUpToDate` condition

Reply:
We should definitely opt for webhook validation in this case, which is immediate and informs the user before reconciliation starts